pilot/pkg/serviceregistry/kube/controller/ambientindex_test.go

```diff
@@ -130,199 +146,244 @@ func TestAmbientIndex(t *testing.T) {
-	assertEvent("127.0.0.4")
+	assertAddresses("", "name1", "name2", "name3", "svc1")
+	assertAddresses("testnetwork/10.0.0.1", "name1", "name2", "svc1")
+	assertAddresses("testnetwork/127.0.0.4") // Should not be accessible anymore
+	assertEvent("testnetwork/127.0.0.4")
 
 	fx.Clear()
 	// Update Service to have a more restrictive label selector
-	createService(controller, "svc1", "ns1",
+	addService("svc1", "ns1",
 		map[string]string{},
-		[]int32{80}, map[string]string{"app": "a", "other": "label"}, t)
-	assertWorkloads("", "name1", "name2", "name3")
-	assertWorkloads("10.0.0.1", "name2")
+		map[string]string{},
+		[]int32{80}, map[string]string{"app": "a", "other": "label"}, "10.0.0.1")
+	assertAddresses("", "name1", "name2", "name3", "svc1")
+	assertAddresses("testnetwork/10.0.0.1", "name2", "svc1")
 	// Need to update the *old* workload only
-	assertEvent("127.0.0.1", "127.0.0.2")
+	assertEvent("ns1/svc1.ns1.svc.cluster.local", "testnetwork/10.0.0.1", "testnetwork/127.0.0.1", "testnetwork/127.0.0.2")
 	// assertEvent("127.0.0.2") TODO: This should be the event, but we are not efficient here.
 
 	// Update an existing pod into the service
 	addPods("127.0.0.3", "name3", "sa1", map[string]string{"app": "a", "other": "label"}, nil)
-	assertWorkloads("", "name1", "name2", "name3")
-	assertWorkloads("10.0.0.1", "name2", "name3")
-	assertEvent("127.0.0.3")
+	assertAddresses("", "name1", "name2", "name3", "svc1")
+	assertAddresses("testnetwork/10.0.0.1", "name2", "name3", "svc1")
+	assertEvent("testnetwork/127.0.0.3")
 
 	// And remove it again
 	addPods("127.0.0.3", "name3", "sa1", map[string]string{"app": "a"}, nil)
-	assertWorkloads("", "name1", "name2", "name3")
-	assertWorkloads("10.0.0.1", "name2")
-	assertEvent("127.0.0.3")
+	assertAddresses("", "name1", "name2", "name3", "svc1")
+	assertAddresses("testnetwork/10.0.0.1", "name2", "svc1")
+	assertEvent("testnetwork/127.0.0.3")
 
 	// Delete the service entirely
-	controller.client.Kube().CoreV1().Services("ns1").Delete(context.Background(), "svc1", metav1.DeleteOptions{})
-	assertWorkloads("", "name1", "name2", "name3")
-	assertWorkloads("10.0.0.1")
-	assertEvent("127.0.0.2")
+	deleteService("svc1")
+	assertAddresses("", "name1", "name2", "name3")
+	assertAddresses("testnetwork/10.0.0.1")
+	assertEvent("ns1/svc1.ns1.svc.cluster.local", "testnetwork/10.0.0.1", "testnetwork/127.0.0.2")
 	assert.Equal(t, len(controller.ambientIndex.byService), 0)
 
-	// Add a waypoint proxy for namespace
-	addPods("127.0.0.200", "waypoint-ns", "namespace-wide", map[string]string{constants.ManagedGatewayLabel: constants.ManagedGatewayMeshControllerLabel}, nil)
-	assertWorkloads("", "name1", "name2", "name3", "waypoint-ns")
+	// Add a waypoint proxy pod for namespace
+	addPods("127.0.0.200", "waypoint-ns", "namespace-wide", map[string]string{constants.ManagedGatewayLabel: constants.ManagedGatewayMeshControllerLabel, constants.GatewayNameLabel: "namespace-wide"}, nil)
```

Comment:
nit: can you name the pod and service differently? so test can make sure they are not mixed up

Reply:
done